SecondStageUpdater/Program.cs

```diff
@@ -197,6 +201,54 @@ private static void Main(string[] args)
         }
     }
 
+    /// <summary>
+    /// This attempts to copy a file for the update with the ability to retry up to <see cref="MaxCopyAttempts"/> times.
+    /// There are instances where DLLs or other files may be locked and are unable to be overwritten by the update.
+    ///
+    /// TODO:
+    /// Make a backup of all files that are attempted. When we check for any failed files outside this function, restore all backups
+    /// if any failures occurred. This will prevent the user from being in a partially updated state.
+    ///
+    /// </summary>
+    /// <param name="baseDirectory">The absolute path of the game installation.</param>
+    /// <param name="fileInfo">The file to be copied.</param>
+    /// <param name="relativeFileInfo">The relative file info for the destination of the file to be copied.</param>
+    /// <param name="failedFiles">If the copy fails too many times, the file should be added to this list.</param>
+    /// <returns>A Task.</returns>
+    private static Task CopyFileTask(DirectoryInfo baseDirectory, FileInfo fileInfo, FileInfo relativeFileInfo, List<FileInfo> failedFiles)
+    {
+        return Task.Run(() =>
+        {
+            for (int attempt = 1; ; attempt++)
+            {
+                try
+                {
+                    FileInfo copiedFile = SafePath.GetFile(baseDirectory.FullName, relativeFileInfo.ToString());
+
+                    fileInfo.CopyTo(copiedFile.FullName, true);

```

Comment:
To get even more out of it:
```suggestion
                FileInfo destinationFile = SafePath.GetFile(baseDirectory.FullName, relativeFileInfo.ToString());
                FileStream sourceFileStream = sourceFileInfo.Open(new FileStreamOptions
                {
                    Access = FileAccess.Read,
                    Mode = FileMode.Open,
                    Options = FileOptions.Asynchronous,
                    Share = FileShare.None
                });

                await using (sourceFileStream.ConfigureAwait(false))
                {
                    FileStream destinationFileStream = destinationFile.Open(new FileStreamOptions
                    {
                        Access = FileAccess.Write,
                        Mode = FileMode.Create,
                        Options = FileOptions.Asynchronous,
                        Share = FileShare.None
                    });

                    await using (destinationFileStream.ConfigureAwait(false))
                    {
                        await sourceFileStream.CopyToAsync(destinationFileStream).ConfigureAwait(false);
                    }
                }

```